src/Umbraco.Cms.Integrations.Search.Algolia/Controllers/SearchController.cs

```diff
@@ -52,61 +52,82 @@ public IActionResult GetIndices()
         } 
 
         [HttpPost]
-        public string SaveIndex([FromBody] IndexConfiguration index)
+        public async Task<IActionResult> SaveIndex([FromBody] IndexConfiguration index)
         {
             try
             {
-                _scopeService.AddOrUpdate(new AlgoliaIndex
+                _indexStorage.AddOrUpdate(new AlgoliaIndex
                 {
                     Id = index.Id,
                     Name = index.Name,
                     SerializedData = JsonSerializer.Serialize(index.ContentData),
                     Date = DateTime.Now
                 });
 
+                var result = await _indexService.PushData(index.Name);
+
+                return new JsonResult(result);
+            }
+            catch(Exception ex)
+            {
+                return new JsonResult(Result.Fail(ex.Message));
+            }
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> BuildIndex([FromBody] IndexConfiguration indexConfiguration)
+        {
+            try
+            {
+                var index = _indexStorage.GetById(indexConfiguration.Id);
+
                 var payload = new List<Record>();
 
-                foreach (var item in index.ContentData)
+                var indexContentData = JsonSerializer.Deserialize<List<ContentData>>(index.SerializedData);
+
+                foreach (var contentDataItem in indexContentData)
                 {
-                    var contentItems = _umbracoHelper.ContentAtXPath($"//{item.ContentType}");
+                    var contentItems = _umbracoHelper.ContentAtXPath($"//{contentDataItem.ContentType}");
```

Comment:
Might be better here to use `IContentService.GetPagedOfTypes()`.  You can do in a loop to not get massive amounts into memory in one go, and think this is a more Umbraco 9+ way to do things now there's no XML cache.